File: src/algorithms/dynamic-programming/string-dp/word-break-tabulation/sources/WordBreakTabulation_test.cpp
// g++ -o test WordBreakTabulation_test.cpp && ./test
#include "WordBreakTabulation.cpp"
#include <cassert>
#include <iostream>
#include <string>
#include <vector>

int main() {
    assert(wordBreakTabulation("leetcode", {"leet", "code"}) == true);
    assert(wordBreakTabulation("applepenapple", {"apple", "pen"}) == true);
    assert(wordBreakTabulation("catsandog", {"cats", "dog", "sand", "and", "cat"}) == false);
    assert(wordBreakTabulation("", {"a"}) == true);
    assert(wordBreakTabulation("catsanddog", {"cats", "dog", "sand", "and", "cat"}) == true);
    assert(wordBreakTabulation("hello", {"world", "foo"}) == false);
    assert(wordBreakTabulation("apple", {"apple", "pen"}) == true);
    assert(wordBreakTabulation("leetcoderr", {"leet", "code"}) == false);

    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: src/algorithms/dynamic-programming/string-dp/word-break-tabulation/__tests__/word-break-tabulation_test.rs
include!("../sources/word-break-tabulation.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn leetcode_returns_true() {
        assert!(word_break_tabulation("leetcode", &["leet", "code"]));
    }

    #[test]
    fn applepenapple_returns_true() {
        assert!(word_break_tabulation("applepenapple", &["apple", "pen"]));
    }

    #[test]
    fn catsandog_returns_false() {
        assert!(!word_break_tabulation("catsandog", &["cats", "dog", "sand", "and", "cat"]));
    }

    #[test]
    fn empty_string_returns_true() {
        assert!(word_break_tabulation("", &["a"]));
    }

    #[test]
    fn catsanddog_returns_true() {
        assert!(word_break_tabulation("catsanddog", &["cats", "dog", "sand", "and", "cat"]));
    }

    #[test]
    fn no_match_returns_false() {
        assert!(!word_break_tabulation("hello", &["world", "foo"]));
    }

    #[test]
    fn exact_match_returns_true() {
        assert!(word_break_tabulation("apple", &["apple", "pen"]));
    }

    #[test]
    fn leftover_returns_false() {
        assert!(!word_break_tabulation("leetcoderr", &["leet", "code"]));
    }
}
